File: Java/Q2AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Q2AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        //initialise a dummy head, current node and carry
        ListNode head = new ListNode(0);
        ListNode curr = head;
        int carry = 0;
        
        //while loop that terminate only if both l1 and l2 is null and carry is 0
        while(l1 != null || l2 != null || carry > 0){
            //computer sum (considering and updating carry)
            int x = (l1 == null? 0: l1.val);
            int y = (l2 == null? 0: l2.val);
            int sum = x+y+carry;
            carry = sum / 10;
            
            //create new node and update current node
            curr.next = new ListNode(sum % 10);
            curr = curr.next;
            
            //advance l1 and l2 if they are not null
            l1 = (l1 == null? null: l1.next);
            l2 = (l2 == null? null: l2.next);
        }
        return head.next;
    }
}

//Runtime: 3 ms, faster than 78.30% of Java online submissions for Add Two Numbers.
//Memory Usage: 47.8 MB, less than 47.23% of Java online submissions for Add Two Numbers.

File: Python/Q2AddTwoNumbers.py
# @b-knd (jingru) on 07 August 2022 15:14:00

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        
        dummyHead = ListNode(0)
        curr = dummyHead
        carry = 0
        while l1 != None or l2 != None or carry > 0:
            x = l1.val if l1 else 0
            y = l2.val if l2 else 0
            currSum = x + y + carry
            newNode = ListNode(currSum % 10)
            carry = currSum // 10
            curr.next = newNode
            curr = curr.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        return dummyHead.next

#Runtime: 84 ms, faster than 54.07% of Python online submissions for Add Two Numbers.
#Memory Usage: 13.6 MB, less than 17.39% of Python online submissions for Add Two Numbers.
